merkle-tree/hash-set/src/lib.rs

```diff
@@ -0,0 +1,1211 @@
+        // SAFETY: `I` is always a signed integer. Creating a layout for an
+        // array of integers of any size won't cause any panic.
+        let values_dst_ptr = unsafe { alloc::alloc(values_layout) as *mut Option<HashSetCell> };
+        if values_dst_ptr.is_null() {
+            handle_alloc_error(values_layout);
+        }
+        let values = NonNull::new(values_dst_ptr).unwrap();
+        for i in 0..capacity_values {
+            std::ptr::write(values_dst_ptr.add(i), None);
+        }
+
+        let offset = offset + indices_size;
+        let values_src_ptr = bytes.as_ptr().add(offset) as *const Option<HashSetCell>;
+        std::ptr::copy(values_src_ptr, values_dst_ptr, capacity_values);
+
+        Ok(Self {
+            capacity_indices,
+            capacity_values,
+            next_value_index,
+            sequence_threshold,
+            indices,
+            values,
+        })
+    }
+
+    fn insert_into_occupied_cell(
+        &mut self,
+        value_index: usize,
+        value: &BigUint,
+        current_sequence_number: usize,
+    ) -> Result<bool, HashSetError> {
+        let value_bucket = unsafe { &mut *self.values.as_ptr().add(value_index) };
+        match value_bucket {
+            // The cell in the value array is already taken.
+            Some(value_bucket) => {
+                // We can overwrite that cell only if the element
+                // is expired - when the difference between its
+                // sequence number and provided sequence number is
+                // greater than the threshold.
+                if let Some(element_sequence_number) = value_bucket.sequence_number {
+                    if current_sequence_number >= element_sequence_number {
+                        *value_bucket = HashSetCell {
+                            value: bigint_to_be_bytes_array(value)?,
+                            sequence_number: None,
+                        };
+                        return Ok(true);
+                    }
+                }
+                // Otherwise, we need to prevent having multiple valid
+                // elements with the same value.
+                if &BigUint::from_be_bytes(value_bucket.value.as_slice()) == value {
+                    return Err(HashSetError::ElementAlreadyExists);
+                }
+            }
+            // Panics: If there is a hash set cell pointing to a `None` value,
+            // it means we really screwed up in the implementation...
+            // That should never happen.
+            None => unreachable!(),
+        }
+        Ok(false)
+    }
+
+    /// Inserts a value into the hash set.
+    pub fn insert(
+        &mut self,
+        value: &BigUint,
+        current_sequence_number: usize,
+    ) -> Result<(), HashSetError> {
+        for i in 0..self.capacity_values {
+            let i = I::try_from(i).map_err(|_| HashSetError::IntegerOverflow)?;
```

Comment:
could we check this in the new method?

Reply:
What do you mean? Moving this `try_from` line somewhere else? I don't think it would change much.

Reply:
Just thought that it is an unnecessary dynamic operation that we loop over all the time  while it could be a one time check during initialization.

Reply:
Now I see what you mean! Yes, I think that converting `self.capacity` values to `I` would address your concern, since we would do the conversion just once and then the whole loop would work on ranges of `I` rather than ranges of `usize`. Let me do that.

Reply:
Or, actually, no, I can't do that :(

After trying that, I got the error

```
error[E0277]: the trait bound `I: Step` is not satisfied
   --> merkle-tree/hash-set/src/lib.rs:433:18
    |
433 |         for i in I::zero()..capacity_values {
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Step` is not implemented for `I`
    |
    = note: required for `std::ops::Range<I>` to implement `Iterator`
    = note: required for `std::ops::Range<I>` to implement `IntoIterator`
help: consider further restricting this bound
    |
198 |         + Unsigned + std::iter::Step,
    |                    +++++++++++++++++
```

and unfortunately, `std::iter::Step` trait is an unstable feature (only available in Nightly) https://doc.rust-lang.org/std/iter/trait.Step.html

Reply:
what's the purpose of the try_from actually?
(when removing the line the hash-set also compiles)